### PR TITLE
Implement custom sorting cli option

### PR DESCRIPTION
**Pull request by hobarrera**
Implement `todo list --sort FIELD1,FIELD2`, to allow custom sorting.

Though the code seems complex, there's no noticeable difference between this sorting mechanism and the original one, though I'd love some feedback on how to de-uglify this.

Code can also be used to make the default sort order configurable.

### CODE REVIEW

File: todoman/main.py [hunk 1]
```diff
@@ -30,25 +30,45 @@ def dump_idfile(ids):
...
+                # Remove that '-'
+                field = field[1:]
+            if field == 'due':
+                # TODO: Some better method to deal with these special cases?
+                value = todo.due.timestamp() if todo.due else float('inf')
+            elif field == 'created_at':
+                value = todo.created_at.timestamp() if todo.created_at else 0
+            else:
+                value = getattr(todo, field, "")
+            value = -value if neg and value else value
```
Comment (untitaker) [inline todoman/main.py]: Haven't tested, but should crash if given `-summary`.

Comment (hobarrera) [inline todoman/main.py]: Yes, it does.

I can't really think of any other way to keep this generic *and* working for al these edge cases.

Comment (untitaker) [inline todoman/main.py]: Fixed!

**Comment by hobarrera** [comment]
Thanks for adding to this, @untitaker. I'll try to add some basic tests that make sure that at least a few pattern works to catch future regressions and merge this into master.

**Comment by untitaker** [comment]
Which tests are missing? Are you talking about ones that actually assert correct output?

**Comment by hobarrera** [comment]
> Which tests are missing? Are you talking about ones that actually assert correct output?

Yes, at least a basic one to catch regression. I've just added one and think that's enough.

**Comment by hobarrera** [comment]
As-is, sorting seems to be the opposite of what one would naturally expect.

Sorting by summary lists:
```
ccc
bbb
aaa
```

Sorting by due date:
```
2015-11-11
2015-06-16
2015-05-31
```

IMHO, we should invert this. Opinions?

**Comment by untitaker** [comment]
Didn't we invert this on purpose such that you don't have to scroll to see the topmost entries?

On 17 April 2016 05:26:54 CEST, Hugo Osvaldo Barrera <notifications@github.com> wrote:
>As-is, sorting seems to be the opposite of what one would naturally
>expect.
>
>Sorting by summary lists:
>```
>ccc
>bbb
>aaa
>```
>
>Sorting by due date:
>```
>2015-11-11
>2015-06-16
>2015-05-31
>```
>
>IMHO, we should invert this. Opinions?
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/todoman/pull/3#issuecomment-210946504

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by untitaker** [comment]
Perhaps we should add a reverse flag in addition, such that you can pipe to a pager and reverse without having to modify your sort key.

On 17 April 2016 05:26:54 CEST, Hugo Osvaldo Barrera <notifications@github.com> wrote:
>As-is, sorting seems to be the opposite of what one would naturally
>expect.
>
>Sorting by summary lists:
>```
>ccc
>bbb
>aaa
>```
>
>Sorting by due date:
>```
>2015-11-11
>2015-06-16
>2015-05-31
>```
>
>IMHO, we should invert this. Opinions?
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/todoman/pull/3#issuecomment-210946504

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by hobarrera** [comment]
I like this idea. It also makes sense for `todo list` in general.

**Comment by untitaker** [comment]
So, `--reverse/--no-reverse` with default=True?

On 17 April 2016 10:32:18 CEST, Hugo Osvaldo Barrera <notifications@github.com> wrote:
>I like this idea. It also makes sense for `todo list` in general.
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/todoman/pull/3#issuecomment-210976225

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by hobarrera** [comment]
> So, `--reverse/--no-reverse` with default=True?

Yes, yes. We'll need to reverse the default order though, so that `reverse=True` does not alter the output of `todo list` (eg: so that it continues printing everything in the same order).

**Comment by untitaker** [comment]
That's what i meant, so you pass --no-reverse if you want to pipe to a pager.

On 17 April 2016 10:38:16 CEST, Hugo Osvaldo Barrera <notifications@github.com> wrote:
>> So, `--reverse/--no-reverse` with default=True?
>
>Yes, yes. We'll need to reverse the default order though, so that
>`reverse=True` does not alter the output of `todo list` (eg: so that it
>continues printing everything in the same order).
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/todoman/pull/3#issuecomment-210977199

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by untitaker** [comment]
I added the --reverse/--no-reverse option with tests.

File: tests/test_basic.py [hunk 2]
```diff
@@ -155,15 +155,30 @@ def test_sorting_output(tmpdir, runner, create):
...
-    result = runner.invoke(cli, ['list', '--sort', '-summary'])
-    assert not result.exception
-    assert 'aaa' in result.output.splitlines()[0]
-    assert 'bbb' in result.output.splitlines()[1]
 
-    result = runner.invoke(cli, ['list', '--sort', 'due'])
-    assert not result.exception
-    assert 'aaa' in result.output.splitlines()[0]
-    assert 'bbb' in result.output.splitlines()[1]
+    examples = [
```
Comment (untitaker) [inline tests/test_basic.py]: This is probably a bit overengineered now, but meh

**Comment by untitaker** [comment]
I have no idea why the tests fail, they work for me outside of tox?

**Comment by hobarrera** [comment]
> I added the --reverse/--no-reverse option with tests.

Thanks. :)

> I have no idea why the tests fail, they work for me outside of tox?

I've no idea - but they pass for me outside tox as well, and am trying to figure out what's going on.

**Comment by untitaker** [comment]
I have the solution, I actually have a massive bug in the tests that creates corrupt todos that are skipped. Combine with broken logging for invalid todos and you have this behavior :)

**Comment by hobarrera** [comment]
> I have the solution, I actually have a massive bug in the tests that creates corrupt todos that are skipped. Combine with broken logging for invalid todos and you have this behavior :)

Since you're using the `Todo` class, I'd expect them to be valid. Maybe also wrap creation using a `Database` class?

**Comment by hobarrera** [comment]
Seems to work fine. I'll do the rebasing/merge once CI finishes, if  you think it's okay.

**Comment by untitaker** [comment]
@hobarrera I'm using `Todo` from icalendar, which seems to behave very funky when doing `todo['value'] = <datetime>`. Fixed to use the todoman one. I'll also add a commit to not directly use icalendar at all.
